substrate/primitives/runtime/src/generic/checked_extrinsic.rs

```diff
@@ -18,81 +18,130 @@
+	General(Extension),
+}
+
+// TODO: Rename ValidateUnsigned to ValidateInherent
+// TODO: Consider changing ValidateInherent API to avoid need for duplicating validate
+//   code into pre_dispatch (rename that to `prepare`).
+// TODO: New extrinsic type corresponding to `ExtrinsicFormat::General`, which is
+//   unsigned but includes extension data.
+// TODO: Move usage of `signed` to `format`:
+// - Inherent instead of None.
+// - Signed(id, extension) instead of Some((id, extra)).
+// - Introduce General(extension) for one without a signature.
+
 /// Definition of something that the external world might want to say; its existence implies that it
 /// has been checked and is good, particularly with regards to the signature.
 ///
 /// This is typically passed into [`traits::Applyable::apply`], which should execute
 /// [`CheckedExtrinsic::function`], alongside all other bits and bobs.
 #[derive(PartialEq, Eq, Clone, sp_core::RuntimeDebug)]
-pub struct CheckedExtrinsic<AccountId, Call, Extra> {
+pub struct CheckedExtrinsic<AccountId, Call, Extension> {
 	/// Who this purports to be from and the number of extrinsics have come before
 	/// from the same signer, if anyone (note this is not a signature).
-	pub signed: Option<(AccountId, Extra)>,
+	pub format: ExtrinsicFormat<AccountId, Extension>,
 
 	/// The function that should be called.
 	pub function: Call,
 }
 
-impl<AccountId, Call, Extra, RuntimeOrigin> traits::Applyable
-	for CheckedExtrinsic<AccountId, Call, Extra>
+impl<AccountId, Call, Extension, RuntimeOrigin> traits::Applyable
+	for CheckedExtrinsic<AccountId, Call, Extension>
 where
 	AccountId: Member + MaybeDisplay,
-	Call: Member + Dispatchable<RuntimeOrigin = RuntimeOrigin>,
-	Extra: SignedExtension<AccountId = AccountId, Call = Call>,
+	Call: Member + Dispatchable<RuntimeOrigin = RuntimeOrigin> + Encode,
+	Extension: TransactionExtension<Call, ()>,
 	RuntimeOrigin: From<Option<AccountId>>,
 {
 	type Call = Call;
 
-	fn validate<U: ValidateUnsigned<Call = Self::Call>>(
+	fn validate<I: ValidateUnsigned<Call = Self::Call>>(
 		&self,
-		// TODO [#5006;ToDr] should source be passed to `SignedExtension`s?
-		// Perhaps a change for 2.0 to avoid breaking too much APIs?
 		source: TransactionSource,
 		info: &DispatchInfoOf<Self::Call>,
 		len: usize,
 	) -> TransactionValidity {
-		if let Some((ref id, ref extra)) = self.signed {
-			Extra::validate(extra, id, &self.function, info, len)
-		} else {
-			let valid = Extra::validate_unsigned(&self.function, info, len)?;
-			let unsigned_validation = U::validate_unsigned(source, &self.function)?;
-			Ok(valid.combine_with(unsigned_validation))
+		match self.format {
+			ExtrinsicFormat::Bare => {
+				let inherent_validation = I::validate_unsigned(source, &self.function)?;
+				#[allow(deprecated)]
+				let legacy_validation = Extension::validate_bare_compat(&self.function, info, len)?;
+				Ok(legacy_validation.combine_with(inherent_validation))
```

Comment:
AFAIK inherents should not be validated. They should only be generated and applied

Reply:
We talked about this offline but didn't post the outcome as it got lost in the 200+ comments, but we will provide the validation with the transaction source.

Reply:
This goes a little bit into the direction of the `IsAuthorized` logic. We need to check that the inherents are actually inherents, otherwise the block producer can just apply any kind of extrinsic. We already had this issue at some point with functions that accepted any kind of origin (like batch).

What was called unsigned transaction before, should be a `General` transaction in the future. There is no difference to the normal general transaction as they just require special validation logic to let them "through" without paying (as before). This brings also up again the point that `RawOrigin::None` is a valid origin as well. 

`BeefyCall::report_double_voting_unsigned` these are not required to be send by block producers and there also exists tons of other similar unsigned transactions.